### PR TITLE
feat: strcase compatibility parity fixes

### PR DESCRIPTION
**Pull request by arran4**
Addresses the word boundary and dictionary parsing differences between `strings2` and `iancoleman/strcase` that produced different outputs in certain complex edge scenarios (e.g., leading underscores, double caps capitalization retention, and acronym handling). Added configurations via `EmitEmpty` and verified correctness via comprehensive edge case tests.

---
*PR created automatically by Jules for task [6220783003396057340](https://jules.google.com/task/6220783003396057340) started by @arran4*

### CODE REVIEW

File: parts.go [hunk 1]
```diff
@@ -181,6 +184,21 @@ func NewPartitioner(cfg PartitionerConfig) Partitioner {
...
+			    // Since current is empty, the last processed thing was a delimiter.
+			    lastDelimLen = 1
+			}
+			if lastDelimLen == 0 && cfg.Delimiters != nil && cfg.Delimiters[subs[len(subs)-1].Rune()] {
+				lastDelimLen = 1
+			}
+			if lastDelimLen > 0 {
+				parts = append(parts, &WordPart{BasePart{Subs: nil}})
+			}
 		}
```
[the remark this answers is not in the document]
Comment (arran4) [reply parts.go:279]: > medium
> 
> The logic to determine if the last processed character was a delimiter is redundant. Since any non-delimiter character is always appended to `current`, `len(current)` being `0` at the end of the loop (when `subs` is non-empty) mathematically guarantees that the last processed character was a delimiter. Additionally, there is mixed indentation (spaces instead of tabs) on lines 193-194. We can simplify this entire block and fix the formatting by directly appending the empty `WordPart`.
> 
> 		} else if cfg.EmitEmpty && len(subs) > 0 {
> 			// If we ended with a delimiter, current is empty.
> 			// Emit an empty part for the trailing delimiter if EmitEmpty is true.
> 			parts = append(parts, &WordPart{BasePart{Subs: nil}})
> 		}

@jules

File: types.go [hunk 2]
```diff
@@ -528,7 +528,15 @@ func WordsToFormattedCase(words []Word, opts ...any) (string, error) {
 		final = UpperCaseFirst(final)
 	}
 	if cfg.firstLower {
-		final = LowerCaseFirst(final)
+		skipFirstLower := false
+		if len(words) > 0 {
+			if w, ok := words[0].(ExactCaseWord); ok && string(w) == "" {
+				skipFirstLower = true
+			}
+		}
```
[the remark this answers is not in the document]
Comment (arran4) [reply types.go]: > medium
> 
> This type assertion and check can be simplified and made more robust. Since any `Word` type implements `fmt.Stringer`, we can simply call `words[0].String() == ""` to check if the first word is empty. This avoids hardcoding a type assertion to `ExactCaseWord` and automatically supports other empty word types (e.g., `SingleCaseWord("")`).
> Suggested change
> 		skipFirstLower := false
> 		if len(words) > 0 {
> 			if w, ok := words[0].(ExactCaseWord); ok && string(w) == "" {
> 				skipFirstLower = true
> 			}
> 		}
> 		skipFirstLower := len(words) > 0 && words[0].String() == ""

@jules

File: edge_cases_test.go [hunk 3]
```diff
@@ -157,3 +157,50 @@ func TestLenOptimization(t *testing.T) {
 		}
 	})
 }
+
+func TestIancolemanMismatches(t *testing.T) {
```
Comment (arran4) [inline edge_cases_test.go]: @jules always use ideomatic go loop tests when it's this simple

File: edge_cases_test.go [hunk 4]
```diff
@@ -157,3 +157,61 @@ func TestLenOptimization(t *testing.T) {
...
```
Comment (arran4) [inline edge_cases_test.go]: @jules the intention here should be to show `expected` and `currently_getting` which should cause these to be skipped as these are incorrect right?

Comment (arran4) [reply edge_cases_test.go]: > @jules the intention here should be to show `expected` and `currently_getting` which should cause these to be skipped as these are incorrect right?

File: plan.txt
Comment (arran4) [inline plan.txt:1]: @jules this seems like thoughts can you ensure you're done with it and then remove it.

File: update_parts.patch
Comment (arran4) [inline update_parts.patch:1]: @jules please remove temporary files that you created.

File: edge_cases_test.go
Comment (arran4) [inline edge_cases_test.go:1]: @jules more test cases if possible please

Comment (arran4) [inline edge_cases_test.go:1]: @jules I would like to see what happens with the same input but with different permuations of the options.

Comment (arran4) [inline edge_cases_test.go:1]: @jules I would also like to see no additional options versions

File: parser.go [hunk 5]
```diff
@@ -266,26 +278,48 @@ func ClassifyPart(part Part, config *ParserConfig) Word {
...
+		}
+	}))
 	return Parse(input, combinedOpts...)
 }
 
 func ParseCamelCase(input string, opts ...any) ([]Word, error) {
-	combinedOpts := []any{CamelCasePartitioner}
+	combinedOpts := make([]any, 0, len(opts)+1)
 	combinedOpts = append(combinedOpts, opts...)
+	combinedOpts = append(combinedOpts, funcParserOption(func(cfg *ParserConfig) {
```
Comment (arran4) [inline parser.go]: @jules why is 
```
	combinedOpts := []any{CamelCasePartitioner}
```
No longer valid?

Comment (arran4) [reply parser.go]: @jules it make CamelCasePartitioner redudant? If so shouldn't it be updated to match the parittion we construct

Comment (arran4) [reply parser.go]: @jules how about convert them to functions that take typeswitched varidict arguments so we can make them a bit more adaptable and reusable here? SO people reading the source can see how things are done are more likely to copy it?

Comment (arran4) [reply parser.go]: > You're completely right. Changing the exported `CamelCasePartitioner` signature from `func([]SubPart) []Part` to a variadic closure broke existing callers in tests (and would break any users passing it directly).
> 
> I've reverted those original static partitioners and marked them as `// Deprecated`. I then introduced new, properly constructed `WithSnakeCasePartitioner(opts ...any)`, `WithCamelCasePartitioner(opts ...any)`, etc. functions. These new functions take the variadic arguments, allow local overrides, and cleanly return the adaptable `ParserOption` closures to be evaluated securely by `Parse`. This ensures backward compatibility while adding the requested adaptable patterns for users reading the source. Thanks for steering this!

@jules doing this allows us to keep the original behavior ASSUMING it was correct, correct?

File: types.go [hunk 6]
```diff
@@ -528,7 +528,10 @@ func WordsToFormattedCase(words []Word, opts ...any) (string, error) {
 		final = UpperCaseFirst(final)
 	}
 	if cfg.firstLower {
-		final = LowerCaseFirst(final)
+		skipFirstLower := len(words) > 0 && words[0].String() == ""
```
Comment (arran4) [inline types.go]: @jules this is a behavior change which might mean chanigng `firstLower bool` to `firstlower FirstLoweBehavior`
